### PR TITLE
Move `SourceMap` initialization

### PR DESCRIPTION
**Pull request by nnethercote**
So it happens at the same time as `SessionGlobals` initialization, rather than shortly afterward.

r? @WaffleLapkin

### CODE REVIEW

**Comment by nnethercote** [comment]
This is a follow-up to #122810, and the first two commits come from that PR. I've put this in a separate PR so we can merge the removal of `CodegenBackend::target_override` first, and see if that causes any problems before making changes that build on that.

File: compiler/rustc_interface/src/tests.rs [hunk 1]
```diff
@@ -25,42 +25,46 @@ use std::num::NonZero;
...
-        target,
-        sysroot,
-        "",
-        None,
-        Arc::default(),
-        Default::default(),
-    );
-    let cfg = parse_cfg(&sess.dcx(), matches.opt_strs("cfg"));
-    (sess, cfg)
+    rustc_span::create_session_globals_then(DEFAULT_EDITION, || {
```
Comment (WaffleLapkin) [inline compiler/rustc_interface/src/tests.rs]: Why not `create_default_session_globals_then`?

Also, am I reading this right that this is test only?

File: compiler/rustc_span/src/lib.rs [hunk 2]
```diff
@@ -104,35 +104,35 @@ pub struct SessionGlobals {
...
-    /// happens after `SessionGlobals`. `set_source_map` does the
-    /// initialization.
-    ///
-    /// This field should only be used in places where the `Session` is truly
-    /// not available, such as `<Span as Debug>::fmt`.
-    source_map: Lock<Option<Lrc<SourceMap>>>,
+    /// The session's source map, if there is one. This field should only be
+    /// used in places where the `Session` is truly not available, such as
+    /// `<Span as Debug>::fmt`.
+    source_map: Option<Lrc<SourceMap>>,
```
Comment (WaffleLapkin) [inline compiler/rustc_span/src/lib.rs:110]: Why is this an option? The previous version of the comment is obviously outdated, so there is another reason now?

File: compiler/rustc_span/src/source_map.rs [hunk 3]
```diff
@@ -167,9 +167,17 @@ struct SourceMapFiles {
     stable_id_to_source_file: UnhashMap<StableSourceFileId, Lrc<SourceFile>>,
 }
 
+/// Used to construct a `SourceMap` with `SourceMap::with_inputs`.
+pub struct SourceMapInputs(
+    pub Box<dyn FileLoader + Send + Sync>,
+    pub FilePathMapping,
+    pub SourceFileHashAlgorithm,
+);
```
Comment (WaffleLapkin) [inline compiler/rustc_span/src/source_map.rs:175]: Would it be better to use a named struct?

**Review by WaffleLapkin — APPROVED** [review]
This mostly looks good. I have a couple comments, but we'll have to wait for #122811 to merge first anyway.

File: compiler/rustc_interface/src/tests.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_interface/src/tests.rs]: It's overwritten by the subsequent commit, but I changed it anyway.

Yes, this is test-only code.

File: compiler/rustc_span/src/lib.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_span/src/lib.rs:110]: There are lots of cases (mostly in tests) where we don't bother setting a source map. That was true previously, but the old comment didn't explain that was the second reason for having `Option`.

File: compiler/rustc_span/src/source_map.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_span/src/source_map.rs:175]: ok

**Comment by nnethercote** [comment]
I have updated and addressed the comments. I think this shouldn't land quite yet, to give a little more time for any fallout from #122810 to happen. I will be around on Monday but then I will be on vacation for three weeks. Maybe this should wait until I get back, just to be safe.

**Comment by WaffleLapkin** [comment]
r=me whenever you feel like we waited enough :)

**Comment by nnethercote** [comment]
There have been no complaints about #122810. I think this is safe to merge.

@bors r=WaffleLapkin

**Pull request merged**